### ISSUE TITLE
tensorflow lite model example

### ISSUE THREAD

**Issue by flutter-painter** [OpenNMT/OpenNMT-tf#1005]
Hi @guillaumekln,

Thank you for a great NL tool,
To check that my android integration of OpenNMT tensorflow lite is correct I'd rather test against a bullet-proof model.
Could you please consider exporting your demo **ende** to lite ?
Else could you please guide me on how to export it ?
The folder structure is slightly different than what I ended up with following documentation.
For example I see a single vocab file .vocab

**Comment by flutter-painter** [comment]
Happy New Year !

I followed the serving documentation to build _ende_ model, it works fine when serving it both in python and tensorflow (docker), but **not in tensorflow lite**.

The saved model works like a charm, for example : 

- Source: how do you do ?
- Target: Wie machst du?


Now the same sentence (how do you do ?) is translated by the tf lite version to gibberish : 

> Landwirtschaft Landwirtschaft zurückzuerstatten zurückzuerstatten zurückzuerstatten Kriegsgerüchten Kriegsgerüchten Kriegsgerüchten Kriegsgerüchten Kriegsgerüchten Kriegsgerüchten Kriegsgerüchten Kriegsgerüchten Kriegsgerüchten drahtlose zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten drahtlose drahtlose drahtlose zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten drahtlose drahtlose drahtlose zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten drahtlose drahtlose drahtlose drahtlose drahtlose drahtlose drahtlose drahtlose drahtlose drahtlose drahtlose drahtlose drahtlose drahtlose drahtlose drahtlose drahtlose zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten Kriegsgerüchten drahtlose Kriegsgerüchten drahtlose drahtlose drahtlose drahtlose drahtlose drahtlose drahtlose drahtlose zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten Kriegsgerüchten Kriegsgerüchten Kriegsgerüchten Kriegsgerüchten Kriegsgerüchten drahtlose drahtlose drahtlose Kriegsgerüchten Kriegsgerüchten Kriegsgerüchten Kriegsgerüchten Kriegsgerüchten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten Kriegsgerüchten Kriegsgerüchten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten zurückzuerstatten Kriegsgerüchten Kriegs

Could you please explain to me why the model hallucinates like this ?

The model weighs 461.4 MB. If needed I can share it.
I exported it to tensorflow lite using :

`onmt-main --config data.yml export --export_dir /home/ende_lite --export_format tflite `

I followed  @gcervantes8 paper from [Extending OpenNMT’s TensorFlow Lite to Include Transformer Models](https://apps.dtic.mil/sti/pdfs/AD1144269.pdf)
I checked and confirm that the initial sentence is mapped correctly to source vocabulary ids (i.e. 705 for "How")
The android demo repo is available here : https://github.com/flutter-painter/vosk-android-demo
 (don't mind the vosk bit, it is used as a stable recent android project)

**Comment by gcervantes8** [comment]
Are you running both models with or without beam search?  Maybe that could explain the difference.

Also a few additional questions, that would be helpful:

- Is this a LSTM model or a Transformer model?  

- Which version of TensorFlow are you using?

I've been able to get the same results from running as a Saved Model vs TFlite before, so I'm curious what would be the difference here.

**Comment by flutter-painter** [comment]
Thank you for your quick reply !

I fetched the ende model mentioned in examples/serving/tensorflow_serving/README.md

`wget https://s3.amazonaws.com/opennmt-models/averaged-ende-export500k-v2.tar.g`

This is why the tf lite model made from it is quite heavy.
I assume it is a transformer, but you should know better.
Note I also faced the same issue of model hallucinating and filling x250 words using my own model trained as a Transformer on a small dataset.

I exported using the basic command 

`onmt-main --config data.yml export --myPath/ende_lite --export_format tflite `

Since I haven't set **beam_width** when exporting, I suppose I don't use Beam Search

I use tensorflow version: 2.13.1

**Comment by flutter-painter** [comment]
Hi @gcervantes8,
May I help in any way in debugging this ? Do you need me to share the model or anything else ?

**Comment by dmar1n** [comment]
Hi,

I'm experiencing a similar issue, which I explained in more detail here: https://forum.opennmt.net/t/wrong-output-from-exported-opennmt-tf-models/5622

In short, I trained a TransformersBigSharedEmbeddings model, which runs as expected. The predictions done during the training look good, and final BLEU scores from validation are above 50 points.

Then, regardless how I convert the model (to CT2, any quantization, or OpenNMT saved model), the model produces predictions similar to @flutter-painter's output. I tried using different checkpoints and also the averaged model, to no avail.

After many tests, I would say the issue is in the export, because when I compare the validation predictions from the training (which are well formed and of good quality) against the predictions of the exported model using the same texts and tokenised in the same way, the resulting translations are still no sense.

As suggested by @gcervantes8, I played with different values of beam, but the result remains very similar. I also tried with other inference parameters, but there is no improvement. 

I have the impression that the exported model has some issue related to the left and right padding, because it seems forced to produce a prediction whose length is equal to the maximum length. For instance, when I use a longer sentence as input, then some part of the prediction looks OK, while still having repeated tokens both at the start and at the end of the sentence.

Versions used:
- OpenNMT-tf 2.32.0
- TensorFlow 2.13.1 
- cuDNN 8.6.0.

I run a quick test with OpenNMT-tf 2.31, TensorFlow 2.11, but I didn't observe any substantial difference.

**Comment by gcervantes8** [comment]
I was able to get it to work with the Ckpt format.  I was having issues converting directly from SavedModel format to TFLite.

My version of TensorFlow was 2.9, I believe it should work with any version greater than 2.4+.
My OpenNMT-tf version is 2.30.0.

I downloaded the checkpoint from [here](https://github.com/OpenNMT/OpenNMT-tf/tree/master/scripts/wmt)
```wget https://s3.amazonaws.com/opennmt-models/averaged-ende-ckpt500k.tar.gz```


I exported with the following command
```
onmt-main --model_type TransformerBase,  --config ~/Downloads/data.yml --model_dir ~/Downloads/averaged-ende-ckpt500k --checkpoint_path ~/Downloads/averaged-ende-ckpt500k/model.ckpt-500000 export --format tflite --output_dir ~/Downloads/tflite-out2
```

Data Config:
```
data:
  source_vocabulary: ~/Downloads/averaged-ende-export500k-v2/assets/wmtende.vocab
  target_vocabulary: ~/Downloads/averaged-ende-export500k-v2/assets/wmtende.vocab
```
Vocabulary files:
I got from the SavedModel export (https://s3.amazonaws.com/opennmt-models/averaged-ende-export500k.tar.gz)

The size of the model from the export is 452.1 MB

The tokens I feed into the TFLite model are
**Tokens**: ▁How ▁do ▁you ▁do
**IDs**: [1866, 153, 63, 153]
**Output**: [600 481  48  34 104   2   0   0 ... 0]
Translated sentence: ▁Wie ▁machen ▁Sie ▁das ?: ▁Wie ▁machen ▁Sie ▁das ?: ▁Wie ▁machen ▁Sie ▁das ?

Let me know if you get any issues in reproducing.

**Comment by flutter-painter** [comment]
Thank you so much for your detailed reply !
I tested and it no longer hallucinates \°/ 
And it is astonishingly fast.
Should we update doc to underline that export only works from checkpoint and not from savemodel ?

If I may fire a last question while I am at it...

In the paper two vocab files are mentioned ("...do this for both files")
What confuse me is that then there is a mention of **a single vocab file** :
//Unknown ID is the same as vocabulary size
I use two vocab files and I am unable to get the _`<unk>`_ word when using :

`int unknownId = vocabSource.size();`

**Comment by gcervantes8** [comment]
That is great to hear!

I usually run the models with Dynamic Range quantization, since they run even faster and I don't see much loss in accuracy.

To your question about <unk> ID.
So in the paper, I referred to using 2 vocabulary files, one for the source language, and one for the target language.

The <unk> id value for the source language will be equal to the vocabulary size of the source language.
And the <unk> id value for the target language will be equal to the vocabulary size of the target language.

If you're using only one vocab file for both the source and target language (like the pretrained en-de model), then the <unk> id value will be equal to the size of that vocab file, for both the source and target language.


My guess as to why your model isn't predicting <unk> would be that it is not encountering <unk> values during training, in which case it won't predict them after the training.
